py-polars/polars/lazyframe/frame.py

```diff
@@ -2003,6 +2003,8 @@ def collect(
         if not (is_config_obj or engine in ("cpu", "gpu")):
             msg = f"Invalid engine argument {engine=}"
             raise ValueError(msg)
+        if get_default_engine() == "gpu":  # pragma: no cover
```

Comment:
I don't think this should dispatch here, but rather in `fn optimize`.

I also think we should do all engines at once. Where we call it "engine_affinity" as we cannot guarantee it will on the preferred engine.

Reply:
OK, so that would be moving the current python-based engine-selection logic into `optimize`? Note how below we need to know here whether or not we have a "gpu" engine to deliver the appropriate `post_opt_callback`, I think.

Reply:
Hmm, right this only works for streaming/in-memory engine. For the `post-opt-callback` this indeed needs to be in python. I will take a better look to better understand where we can have that logic.

Reply:
Hey @ritchie46, I appreciate the feedback so far. I've addressed the changes except for fully moving the engine-selection logic into Rust. Given that piece might take a bit longer to finalize (especially how we handle the `post_opt_callback`), would you be open to merging this PR as-is? 

I'm totally happy to keep iterating on the Rust-based dispatch in a follow-up PR. WDYT?